File: leetcode-submissions/0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> uniques = new HashMap<>();
        char[] chars = s.toCharArray();
        int left = 0;
        int ans = 0;
        for(int right = 0; right < chars.length; right++) {
            char c = chars[right];            
            if(uniques.containsKey(c)) {
                ans = Math.max(ans, right - left); 
                int newLeft = uniques.get(c) + 1;
                while(left < newLeft) { // should update the ones we skipped... sliding window
                    uniques.remove(chars[left]);
                    left++;
                }
            } else {
                ans = Math.max(ans, right - left + 1);
            }
            uniques.put(c, right);
        } 
        return ans;
    }
}

File: leetcode-submissions/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        dups = set()
        ans = 0
        for r in s:
            while r in dups:
                dups.remove(s[l])
                l += 1
            dups.add(r)
            ans = max(len(dups), ans)
        return ans

